etcd/etcd.go

```diff
@@ -0,0 +1,207 @@
+	if err != nil {
+		return nil, err
+	}
+	c := &client{
+		ecli:               etcdClient,
+		parser:             opts.ConfigParser,
+		prefixTemplate:     prefixTemplate,
+		serverPathTemplate: serverNameTemplate,
+		clientPathTemplate: clientNameTemplate,
+	}
+	return c, nil
+}
+
+func (c *client) SetParser(parser ConfigParser) {
+	c.parser = parser
+}
+
+func (c *client) ClientConfigParam(cpc *ConfigParamConfig, cfs ...CustomFunction) (Key, error) {
+	return c.configParam(cpc, c.clientPathTemplate, cfs...)
+}
+
+func (c *client) ServerConfigParam(cpc *ConfigParamConfig, cfs ...CustomFunction) (Key, error) {
+	return c.configParam(cpc, c.serverPathTemplate, cfs...)
+}
+
+// configParam render config parameters. All the parameters can be customized with CustomFunction.
+// ConfigParam explain:
+//  1. Prefix: KitexConfig by default.
+//  2. ServerPath: {{.ServerServiceName}}/{{.Category}} by default.
+//     ClientPath: {{.ClientServiceName}}/{{.ServerServiceName}}/{{.Category}} by default.
+func (c *client) configParam(cpc *ConfigParamConfig, t *template.Template, cfs ...CustomFunction) (Key, error) {
+	param := Key{}
+
+	var err error
+	param.Path, err = c.render(cpc, t)
+	if err != nil {
+		return param, err
+	}
+	param.Prefix, err = c.render(cpc, c.prefixTemplate)
+	if err != nil {
+		return param, err
+	}
+
+	for _, cf := range cfs {
+		cf(&param)
+	}
+	return param, nil
+}
+
+func (c *client) render(cpc *ConfigParamConfig, t *template.Template) (string, error) {
+	var tpl bytes.Buffer
+	err := t.Execute(&tpl, cpc)
+	if err != nil {
+		return "", err
+	}
+	return tpl.String(), nil
+}
+
+// RegisterConfigCallback register the callback function to etcd client.
+func (c *client) RegisterConfigCallback(ctx context.Context, key string, clientKey int64, callback func(string, ConfigParser)) {
+	clientCtx, cancel := context.WithCancel(context.Background())
+	go func() {
+		m.Lock()
+		tmp := key + "/" + strconv.FormatInt(clientKey, 10)
+		ctxMap[tmp] = cancel
+		m.Unlock()
+		watchChan := c.ecli.Watch(ctx, key)
+		for {
+			select {
+			case <-clientCtx.Done():
+				return
+			case watchResp := <-watchChan:
+				for _, event := range watchResp.Events {
+					eventType := mvccpb.Event_EventType(event.Type)
+					// 检查事件类型
+					if eventType == mvccpb.PUT {
+						// 配置被更新
+						value := string(event.Kv.Value)
+						klog.Debugf("[etcd] config key: %s updated,value is %s", key, value)
+						callback(value, c.parser)
+					} else if eventType == mvccpb.DELETE {
+						// 配置被删除
+						klog.Debugf("[etcd] config key: %s deleted", key)
+						callback("", c.parser)
+					}
+				}
+			}
+		}
+	}()
+
+	data, err := c.ecli.Get(context.Background(), key)
+	// the etcd client has handled the not exist error.
+	if err != nil {
+		klog.Debugf("[etcd] key: %s config get value failed", key)
+		return
+	}
+
+	callback(string(data.Kvs[0].Value), c.parser)
+}
+
+func (c *client) DeregisterConfig(key string, clientKey int64) {
```

Comment:
The key question is when should `DeregisterConfig` be called.

For kitex server, `server.RegisterShutdownHook` is a good choise, and for kitex client, it could be used with `client.WithCloseCallback`